presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/PrecisionRecallStateFactory.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.state;
+
+import com.facebook.presto.array.ObjectBigArray;
+import com.facebook.presto.operator.aggregation.FixedNumericHistogram;
+import com.facebook.presto.spi.function.AccumulatorStateFactory;
+
+import static java.util.Objects.requireNonNull;
+
+public class PrecisionRecallStateFactory
+        implements AccumulatorStateFactory<PrecisionRecallState>
+{
+    @Override
+    public PrecisionRecallState createSingleState()
+    {
+        return new SingleState();
+    }
+
+    @Override
+    public Class<? extends PrecisionRecallState> getSingleStateClass()
+    {
+        return SingleState.class;
+    }
+
+    @Override
+    public PrecisionRecallState createGroupedState()
+    {
+        return new GroupedState();
+    }
+
+    @Override
+    public Class<? extends PrecisionRecallState> getGroupedStateClass()
+    {
+        return GroupedState.class;
+    }
+
+    public static class GroupedState
+            extends AbstractGroupedAccumulatorState
+            implements PrecisionRecallState
+    {
+        private boolean nullSet;
+        private final ObjectBigArray<FixedNumericHistogram> trueWeights = new ObjectBigArray<>();
```

Comment:
This is optional, and somewhat lengthy.  I leave it up to you if you want to do this.

Presto, as you know, holds everything in memory, and won't spill for aggregations.  It also runs within the JVM and so it's sensitive to GC latency.  In the past, some aggregation functions which use `ObjectBigArray` of individual states (per group) have caused issues (the groupings can be in the billions).  Some aggregation functions have been rewritten in such a way that their state for grouped aggregation is totally flat--meaning, there's a constant order of objects created to represent the grouped state, not order of N.  This is done to make them more GC-friendly.  Typically, this involves stuffing the group id into the state itself somehow (for example, `array_agg` does this by interleaving the individual array aggregations via a linked list, which we then traverse over to reconstruct the per-grouping array aggregation).

It should be relatively easy to do something similar for this, because we know exactly how big each histogram is.  So we could have some sort of GroupedHistogram, which takes in the group id plus the value, and we use the group id to find the starting index into a region of an underlying giant (and GC-friendly) array that holds the histogram state for all groups.

It may be a bit of a premature optimization, because I'm not sure how in practice this function may be used, and how large the histograms could grow.  So this is optional feedback.

CC: @wenleix

Reply:
Very interesting point. For precision/recall, I think it is less relevant (esp. before evidence to the contrary). Meaningful precision/recall is typically done by training over large numbers of instances, so the case where there is a large number of partitions, with meaningful precision/recall for each, seems to me less common. Perhaps we could do this PR without it, and then consider?